Mock Groovy Object expecteding "get" rather than "is" with dot notation accessor

# Issue description

In our testing of a groovy / grails application we have an object under test, into which we inject a Mock of an object, to mock out a method that is called.  In our migration to spock 2 (org.spockframework:spock-core:2.0-M4-groovy-3.0) we have hit an issue when the mock object has a method that returns a boolean - the method cannot be found when accessed by groovy's dot notation accessor.

When the dot notation is used for a boolean it looks for a method with "get" as the method prefix rather than "is". I have recreated this in a standalone groovy script (which I will link to).  So the Mock is an mock of a real object, which the class is:

```groovy
class ExampleData {

    boolean isCurrent() {
        true
    }
}
```

When this ExampleData is created as a Mock, its then called in the controller as follows:

```groovy
    def show1() {
       data.current ? "Data is current" : "Data is not current"
    }
```

When this is called - we get the following error:

```
JUnit5 launcher: passed=1, failed=1, skipped=0, time=502ms

Failures (1):
  Spock:TestV3:Test Mock Object Boolean accessor via dot-notation
    MethodSource [className = 'TestV3', methodName = 'Test Mock Object Boolean accessor via dot-notation', methodParameterTypes = '']
    => groovy.lang.MissingMethodException: No signature of method: ExampleData.getCurrent() is applicable for argument types: () values: []
Possible solutions: isCurrent()
       ExampleController.show1(testV3.groovy:61)
       TestV3.Test Mock Object Boolean accessor via dot-notation(testV3.groovy:23)
Caught: groovy.lang.GroovyRuntimeException
groovy.lang.GroovyRuntimeException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
Caused by: groovy.lang.MissingMethodException: No signature of method: ExampleData.getCurrent() is applicable for argument types: () values: []
Possible solutions: isCurrent()
	at ExampleController.show1(testV3.groovy:61)
	at TestV3.Test Mock Object Boolean accessor via dot-notation(testV3.groovy:23)
```

So you can see when accessed via dot notation then you can see that its looking for `getCurrent()` rather than `isCurrent()`

# How to reproduce

So this can be reproduced with the following groovy code:

```groovy
@Grab('cglib:cglib-nodep:3.3.0')
@Grab('org.spockframework:spock-core:2.0-M4-groovy-3.0')
@GrabExclude('org.codehaus.groovy:*')
import spock.lang.Shared
import spock.lang.Specification


class TestV3 extends Specification {

    @Shared 
    ExampleData mockData

    def setup() {
        mockData = Mock(ExampleData) 
    }

    def "Test Mock Object Boolean accessor via dot-notation" () {

        given: "Test Controller"
        ExampleController controller = new ExampleController(data: mockData)

        when: "call test method"
        def result = controller.show1()

        then: "should call mock data"
        1 * mockData.isCurrent() >> true

        and: "should equal"
        result == "Data is current"
    }

    def "Test Mock Object Boolean accessor via method" () {

        given: "Test Controller"
        ExampleController controller = new ExampleController(data: mockData)

        when: "call test method"
        def result = controller.show2()

        then: "should call mock data"
        1 * mockData.isCurrent() >> true

        and: "should equal"
        result == "Data is current"
    }
}

class ExampleData {

    boolean isCurrent() {
        true
    }
}


class ExampleController {

    ExampleData data

    def show1() {
       data.current ? "Data is current" : "Data is not current" 
    }

    def show2() {
       data.isCurrent() ? "Data is current" : "Data is not current" 
    }

}

```

To run:

```
groovy testV3
```

You will see the error as per the description.

I can show that this works with previous version of spock by changing the dependency:

```
 diff testV2.groovy testV3.groovy
2c2
< @Grab('org.spockframework:spock-core:1.3-groovy-2.5')
---
> @Grab('org.spockframework:spock-core:2.0-M4-groovy-3.0')
8c8
< class TestV2 extends Specification {
---
> class TestV3 extends Specification {
```

and running with groovy v2.  This works

# Additional Environment information

## Java/JDK

```
openjdk version "11.0.8" 2020-07-14 LTS
OpenJDK Runtime Environment Zulu11.41+23-CA (build 11.0.8+10-LTS)
OpenJDK 64-Bit Server VM Zulu11.41+23-CA (build 11.0.8+10-LTS, mixed mode)
```

## Groovy version

```
Groovy Version: 3.0.4 JVM: 11.0.8 Vendor: Azul Systems, Inc. OS: Mac OS X
```

## Build tool version

Grapes

## Operating System

Mac OS X 10.15.5

## Build-tool dependencies used

### Groovy Grape

```groovy
@Grab('cglib:cglib-nodep:3.3.0')
@Grab('org.spockframework:spock-core:2.0-M4-groovy-3.0')
@GrabExclude('org.codehaus.groovy:*')
```

--- a/spock-core/src/main/java/org/spockframework/mock/runtime/JavaMockInterceptor.java
+++ b/spock-core/src/main/java/org/spockframework/mock/runtime/JavaMockInterceptor.java
@@ -85,7 +85,7 @@ public class JavaMockInterceptor implements IProxyBasedMockInterceptor {
           if (metaMethod == null) {
             MetaMethod booleanVariant = metaClass
               .getMetaMethod(GroovyRuntimeUtil.propertyToMethodName("is", propertyName), GroovyRuntimeUtil.EMPTY_ARGUMENTS);
-            if (booleanVariant.getReturnType() == boolean.class) {
+            if (booleanVariant != null && booleanVariant.getReturnType() == boolean.class) {
               methodName = booleanVariant.getName();
             }
           }

--- a/spock-core/src/main/java/org/spockframework/mock/runtime/JavaMockInterceptor.java
+++ b/spock-core/src/main/java/org/spockframework/mock/runtime/JavaMockInterceptor.java
@@ -38,7 +38,7 @@ public class JavaMockInterceptor implements IProxyBasedMockInterceptor {
   @Override
   public Object intercept(Object target, Method method, Object[] arguments, IResponseGenerator realMethodInvoker) {
     IMockObject mockObject = new MockObject(mockConfiguration.getName(), mockConfiguration.getExactType(),
-        target, mockConfiguration.isVerified(), false, mockConfiguration.getDefaultResponse(), specification, this);
+      target, mockConfiguration.isVerified(), false, mockConfiguration.getDefaultResponse(), specification, this);
 
     if (method.getDeclaringClass() == ISpockMockObject.class) {
       return mockObject;
@@ -78,7 +78,17 @@ public class JavaMockInterceptor implements IProxyBasedMockInterceptor {
         if (!(mockCaller.getClassName().startsWith("groovy.lang.GroovyObject$getProperty") && "call".equals(mockCaller.getMethodName()))) {
           //HACK: Only explicit getter executions (go.foo and go.getFoo()) should be deeper processed.
           //go.getProperty("foo") is treated as is (to allow for its stubbing)
-          String methodName = GroovyRuntimeUtil.propertyToMethodName("get", (String)args[0]);
+          String propertyName = (String)args[0];
+          MetaClass metaClass = ((GroovyObject)target).getMetaClass();
+          String methodName = GroovyRuntimeUtil.propertyToMethodName("get", propertyName);
+          MetaMethod metaMethod = metaClass.getMetaMethod(methodName, GroovyRuntimeUtil.EMPTY_ARGUMENTS);
+          if (metaMethod == null) {
+            MetaMethod booleanVariant = metaClass
+              .getMetaMethod(GroovyRuntimeUtil.propertyToMethodName("is", propertyName), GroovyRuntimeUtil.EMPTY_ARGUMENTS);
+            if (booleanVariant.getReturnType() == boolean.class) {
+              methodName = booleanVariant.getName();
+            }
+          }
           return GroovyRuntimeUtil.invokeMethod(target, methodName);
         }
       }
@@ -87,7 +97,7 @@ public class JavaMockInterceptor implements IProxyBasedMockInterceptor {
     IMockMethod mockMethod = new StaticMockMethod(method, mockConfiguration.getExactType());
     IMockInvocation invocation = new MockInvocation(mockObject, mockMethod, Arrays.asList(args), realMethodInvoker);
     IMockController mockController = specification == null ? getFallbackMockController() :
-                                                             specification.getSpecificationContext().getMockController();
+      specification.getSpecificationContext().getMockController();
 
     return mockController.handle(invocation);
   }
@@ -96,22 +106,21 @@ public class JavaMockInterceptor implements IProxyBasedMockInterceptor {
     return method.getName().equals(name) && Arrays.equals(method.getParameterTypes(), parameterTypes);
   }
 
-	@Override
+  @Override
   public void attach(Specification specification) {
-		this.specification = specification;
+    this.specification = specification;
 
-	}
+  }
 
-	@Override
+  @Override
   public void detach() {
-	  this.specification = null;
-	}
+    this.specification = null;
+  }
 
-	public MockController getFallbackMockController() {
-	  if (fallbackMockController == null) {
-	    fallbackMockController = new MockController();
-	  }
+  public MockController getFallbackMockController() {
+    if (fallbackMockController == null) {
+      fallbackMockController = new MockController();
+    }
     return fallbackMockController;
   }
 }
-

--- a/spock-specs/src/test/groovy/org/spockframework/smoke/mock/JavaMocksForGroovyClasses.groovy
+++ b/spock-specs/src/test/groovy/org/spockframework/smoke/mock/JavaMocksForGroovyClasses.groovy
@@ -242,7 +242,6 @@ class JavaMocksForGroovyClasses extends Specification {
 
   @Issue("https://github.com/spockframework/spock/issues/1256")
   def "Mock object boolean (is) accessor via dot-notation" () {
-
     given:
     ExampleData mockData = Mock(ExampleData)
 
@@ -258,7 +257,6 @@ class JavaMocksForGroovyClasses extends Specification {
 
   @Issue("https://github.com/spockframework/spock/issues/1256")
   def "Mock object boolean (get) accessor via dot-notation" () {
-
     given:
     ExampleData mockData = Mock(ExampleData)
 
@@ -273,8 +271,51 @@ class JavaMocksForGroovyClasses extends Specification {
   }
 
   @Issue("https://github.com/spockframework/spock/issues/1256")
-  def "Mock object boolean accessor via method" () {
+  def "Mock object boolean (get + is) accessor via dot-notation" () {
+    given:
+    ExampleData mockData = Mock(ExampleData)
 
+    when: "query via property syntax"
+    def result = mockData.active ? "Data is current" : "Data is not current"
+
+    then: "calls mock, preferring 'get' to 'is' for boolean getters (surprise!)"
+    1 * mockData.getActive() >> true
+    0 * mockData.isActive()
+
+    and:
+    result == "Data is current"
+  }
+
+  @Issue("https://github.com/spockframework/spock/issues/1256")
+  def "Mock object non-boolean (get + is) accessor via dot-notation" () {
+    given:
+    ExampleData mockData = Mock(ExampleData)
+
+    when: "query via property syntax"
+    def result = mockData.name ? "Data is current" : "Data is not current"
+
+    then: "calls mock, preferring 'get' to 'is' for non-boolean getters"
+    1 * mockData.getName() >> "X"
+    0 * mockData.isName()
+
+    and:
+    result == "Data is current"
+  }
+
+  @Issue("https://github.com/spockframework/spock/issues/1256")
+  def "Mock object non-boolean (is) pseudo accessor via dot-notation" () {
+    given:
+    ExampleData mockData = Mock(ExampleData)
+
+    when: "query via property syntax"
+    mockData.dummy
+
+    then: "non-boolean 'is' getter is illegal"
+    thrown MissingPropertyException
+  }
+
+  @Issue("https://github.com/spockframework/spock/issues/1256")
+  def "Mock object boolean accessor via method" () {
     given:
     ExampleData mockData = Mock(ExampleData)
 
@@ -316,12 +357,31 @@ class GMock extends AGMock {
 }
 
 class ExampleData {
-
   boolean isCurrent() {
     false
   }
 
   boolean getEnabled() {
     false
+  }
+
+  boolean isActive() {
+    false
+  }
+
+  boolean getActive() {
+    false
+  }
+
+  String isName() {
+    null
+  }
+
+  String getName() {
+    null
+  }
+
+  String isDummy() {
+    null
   }
 }

--- a/spock-specs/src/test/groovy/org/spockframework/smoke/mock/JavaMocksForGroovyClasses.groovy
+++ b/spock-specs/src/test/groovy/org/spockframework/smoke/mock/JavaMocksForGroovyClasses.groovy
@@ -311,7 +311,8 @@ class JavaMocksForGroovyClasses extends Specification {
     mockData.dummy
 
     then: "non-boolean 'is' getter is illegal"
-    thrown MissingPropertyException
+    GroovyRuntimeException exception = thrown()
+    exception instanceof MissingPropertyException || exception instanceof MissingMethodException
   }
 
   @Issue("https://github.com/spockframework/spock/issues/1256")

--- a/spock-specs/src/test/groovy/org/spockframework/smoke/mock/JavaMocksForGroovyClasses.groovy
+++ b/spock-specs/src/test/groovy/org/spockframework/smoke/mock/JavaMocksForGroovyClasses.groovy
@@ -239,39 +239,53 @@ class JavaMocksForGroovyClasses extends Specification {
     then:
     1 * sut.getFoo()
   }
-  
+
   @Issue("https://github.com/spockframework/spock/issues/1256")
-  def "Test Mock Object Boolean accessor via dot-notation" () {
+  def "Mock object boolean (is) accessor via dot-notation" () {
 
-    given: "Test Controller"
+    given:
     ExampleData mockData = Mock(ExampleData)
-    ExampleController controller = new ExampleController(data: mockData)
 
-    when: "call test method"
-    def result = controller.show1()
+    when: "query via property syntax"
+    def result = mockData.current ? "Data is current" : "Data is not current"
 
-    then: "should call mock data"
+    then: "calls mock"
     1 * mockData.isCurrent() >> true
 
-    and: "should equal"
+    and:
     result == "Data is current"
   }
 
   @Issue("https://github.com/spockframework/spock/issues/1256")
-  def "Test Mock Object Boolean accessor via method" () {
+  def "Mock object boolean (get) accessor via dot-notation" () {
 
-    given: "Test Controller"
+    given:
     ExampleData mockData = Mock(ExampleData)
-    ExampleController controller = new ExampleController(data: mockData)
 
-    when: "call test method"
-    def result = controller.show2()
+    when: "query via property syntax"
+    def result = mockData.enabled ? "Data is current" : "Data is not current"
 
-    then: "should call mock data"
+    then: "calls mock"
+    1 * mockData.getEnabled() >> true
+
+    and:
+    result == "Data is current"
+  }
+
+  @Issue("https://github.com/spockframework/spock/issues/1256")
+  def "Mock object boolean accessor via method" () {
+
+    given:
+    ExampleData mockData = Mock(ExampleData)
+
+    when: "query via method syntax"
+    def result = mockData.isCurrent() ? "is enabled" : "is not enabled"
+
+    then: "calls mock"
     1 * mockData.isCurrent() >> true
 
-    and: "should equal"
-    result == "Data is current"
+    and:
+    result == "is enabled"
   }
 }
 
@@ -304,19 +318,10 @@ class GMock extends AGMock {
 class ExampleData {
 
   boolean isCurrent() {
-    true
-  }
-}
-
-
-class ExampleController {
-  ExampleData data
-
-  def show1() {
-    data.current ? "Data is current" : "Data is not current"
+    false
   }
 
-  def show2() {
-    data.isCurrent() ? "Data is current" : "Data is not current"
+  boolean getEnabled() {
+    false
   }
 }

--- a/spock-specs/src/test/groovy/org/spockframework/smoke/mock/JavaMocksForGroovyClasses.groovy
+++ b/spock-specs/src/test/groovy/org/spockframework/smoke/mock/JavaMocksForGroovyClasses.groovy
@@ -239,6 +239,40 @@ class JavaMocksForGroovyClasses extends Specification {
     then:
     1 * sut.getFoo()
   }
+  
+  @Issue("https://github.com/spockframework/spock/issues/1256")
+  def "Test Mock Object Boolean accessor via dot-notation" () {
+
+    given: "Test Controller"
+    ExampleData mockData = Mock(ExampleData)
+    ExampleController controller = new ExampleController(data: mockData)
+
+    when: "call test method"
+    def result = controller.show1()
+
+    then: "should call mock data"
+    1 * mockData.isCurrent() >> true
+
+    and: "should equal"
+    result == "Data is current"
+  }
+
+  @Issue("https://github.com/spockframework/spock/issues/1256")
+  def "Test Mock Object Boolean accessor via method" () {
+
+    given: "Test Controller"
+    ExampleData mockData = Mock(ExampleData)
+    ExampleController controller = new ExampleController(data: mockData)
+
+    when: "call test method"
+    def result = controller.show2()
+
+    then: "should call mock data"
+    1 * mockData.isCurrent() >> true
+
+    and: "should equal"
+    result == "Data is current"
+  }
 }
 
 class MockMe {
@@ -264,5 +298,25 @@ class GMock extends AGMock {
 
   void setFoo(String val) {
     this.prop = val
+  }
+}
+
+class ExampleData {
+
+  boolean isCurrent() {
+    true
+  }
+}
+
+
+class ExampleController {
+  ExampleData data
+
+  def show1() {
+    data.current ? "Data is current" : "Data is not current"
+  }
+
+  def show2() {
+    data.isCurrent() ? "Data is current" : "Data is not current"
   }
 }

